BUG: fixes bug when using sep=None and comment keyword for read_csv

s-scherrer opened the pull request:
- [x] closes #31396 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry - added

s-scherrer committed:
TST: Added test for sep=None and comment='#'

Added a test case to reproduce issue #31396.

s-scherrer committed:
BUG: read_csv: sep=None and comment!=None now possible

This makes read_csv work when sep=None and comment is set to a value.
Fixes #31396.

jreback commented on pandas/tests/io/parser/test_python_parser_only.py:
```diff
@@ -66,6 +66,23 @@ def test_sniff_delimiter(python_parser_only, kwargs):
     tm.assert_frame_equal(result, expected)
 
 
+def test_sniff_delimiter_comment(python_parser_only):
```
can you move this near other comment tests. & add the issue number as a comment

s-scherrer replied:
I moved the test to the end of the file, is this what you meant? Or would you rather have me put it in tests/io/parser/test_comment.py?

gfyoung replied:
`test_comment.py`

s-scherrer replied:
Should I move the full test or simply add `dict(sep=None)` as parameter in `test_line_comment` and then skip the c-enginge tests like it's done for the lineterminator test?

gfyoung replied:
Move the full test as is and change the fixture from `python_parser_only` to `all_parsers`.

s-scherrer replied:
This issue affects only `PythonParser`, `sep=None` does not work with the c engine.

gfyoung replied:
Oh, I see...

Actually, never mind then.  Per the directive in this file, this test is purely for Python parser-only tests, with the goal of migrating them out of here at later date.

Let's keep the test here in fact.

jreback reviewed (changes requested):
can you add a whatsnew note. let's target 1.1 with this.

s-scherrer committed:
DOC: Added whatsnew note and moved test

Added a note in whatsnew/v1.0.0.rst and moved test for #31396 to the end
of tests/io/parser/test_python_parser_only.py.

s-scherrer committed:
DOC: added issue number to whatsnew note

gfyoung commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -180,7 +180,7 @@ I/O
 - `read_csv` will now raise a ``ValueError`` when the arguments `header` and `prefix` both are not `None`. (:issue:`27394`)
 - Bug in :meth:`DataFrame.to_json` was raising ``NotFoundError`` when ``path_or_buf`` was an S3 URI (:issue:`28375`)
--
+- Bug in :meth:`read_csv` was raising `TypeError` when `sep=None` was used in combination with `comment` keyword (:issue:`31396`)
```
Please resolve the merge conflict.

s-scherrer committed:
Merge branch 'master' into sep-none-and-comment

s-scherrer commented:
When adding `dict(sep=None)` as parameter to `test_line_comment` I found that my current fix does not work if the first line is a comment.

Overall I think the flow of comment removal and separating columns is a bit inconsistent. During searching for the bug I found that `PythonParser._check_comments` is sometimes called before separating the line (e.g. when sniffing the separator) and sometimes afterwards. In my opinion it would make more sense  to remove comments first in all cases.

s-scherrer committed:
BUG: sep=None and comment works also when first line is a comment

Deleted old test case and added test setup as parameter to
`test_line_comment`.
The new test showed that the previous fix did not work in case the
first line was a comment line. This commit should fix this.

s-scherrer committed:
TST: Moved test back to test_python_parser_only.py

s-scherrer commented:
The Travis CI build failed with the following message:
```
UnavailableInvalidChannel: The channel is not accessible or is invalid.

  channel name: c3i_test

  channel url: https://conda.anaconda.org/c3i_test

  error code: 404
```

Is there a way to rerun it?

gfyoung commented:
@s-scherrer : This is actually an unrelated error that we need to resolve on `master`.

cc @pandas-dev/pandas-core

s-scherrer committed:
Merge branch 'master' into sep-none-and-comment

WillAyd reviewed (approved):
lgtm

jreback commented:
thanks @s-scherrer
